File: CS-Examples/02_SMTP Client/SendMail.cs
using System;
using System.Windows.Forms;
using OAuth;
using Spire.Email;
using Spire.Email.Smtp;

namespace SendMail
{
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();
        }
        private void btnRun_Click(object sender, EventArgs e)
        {
            if(rBSync.Checked)
            {
                SendEmailSynchronously();
            }
            else if(rBAsync.Checked)
            {
                SendEmailAsynchronously();
            }
        }
        private SmtpClient GetSmtpClient()
        {
            //Connect outlook-Office using OAuth2.0 authentication
            /*   
             *Please use TargetFrameworkVersion 4.5 or above to run this demo
           
             *Dependencies
             AzureROPCTokenProvider.cs
             AzureTokenResponse.cs
             OAuthToken.cs
             TokenType.cs
             .\packages\Newtonsoft.Json.13.0.3\lib\net40\Newtonsoft.Json.dll
             */
			 
            string host = this.HostInput.Text;
            string userName = this.Username.Text;
            string password = this.Password.Text;
            ushort port = ushort.Parse(this.PortInput.Text);

            //Register your application with Azure AD to use Microsoft/Office365 OAUTH in your application, and getting the tenantId and clientId from the registered application.
            string tenantId = "9facb510-bbc6-4c95-a0f5-9a60d3315f3d";
            string clientId = "0a7976ae-ca80-4cb8-983d-9fea41e88dac";

            //scope: The scope to which the client is limited. Permissions for openid, offline_access and https://outlook.office.com/SMTP.Send.         
            string[] scopes = new string[]

            { "openid", "email", "offline_access","https://outlook.office.com/SMTP.Send"};

            //Creating AzureROPCTokenProvider object to get an access token from a token server.
            AzureROPCTokenProvider provider = new AzureROPCTokenProvider(tenantId, clientId, userName, password, scopes);
            OAuthToken result = provider.GetAccessToken();

            //Creating SmtpClient object to connect Outlook server
            SmtpClient client = new SmtpClient(host, port, true, userName, result.Token, ConnectionProtocols.StartTls);
            client.Password = password;

            return client;


            //Do not use OAuth2.0 authentication to connect to qq-email or Gmail
            /*SmtpClient client = new SmtpClient();
            client.Host = this.HostInput.Text;
            client.Username = this.Username.Text ;
            client.Password = this.Password.Text ;
            client.Port = int.Parse(this.PortInput.Text);
            client.ConnectionProtocols = ConnectionProtocols.Ssl;
            return client;*/
        }
        private void SendEmailSynchronously()
        {
            MailMessage msg = new MailMessage(this.FromAddress.Text,this.ToAddress.Text);
            setContent(msg);
            SmtpClient client = GetSmtpClient();
            try
            {
                client.SendOne(msg);
                MessageBox.Show("Sent Successfully");
            }
            catch (Exception ex)
            {
                MessageBox.Show(ex.Message);
            }

        }
        private void SendEmailAsynchronously()
        {
            MailMessage msg = new MailMessage(new MailAddress(this.FromAddress.Text,this.DisplayName.Text),this.ToAddress.Text);
            setContent(msg);
            SmtpClient client = GetSmtpClient();
            client.BeginSendOne(msg, Callback);
            
        }
        private void setContent(MailMessage msg)
        {
            msg.Cc.Add(this.CcInput.Text);
            msg.Bcc.Add(this.BccInput.Text);
            msg.Subject = this.SubjectInput.Text;
            msg.BodyText = this.TextInput.Text;
        }
        AsyncCallback Callback = delegate(IAsyncResult ar)
        {
            IAsyncResultExt task = ar as IAsyncResultExt;
            if (task != null && task.IsCanceled)
            {
                MessageBox.Show("Cancel");
            }
            if (task != null && task.errorInfor != null)
            {
                MessageBox.Show("{0}", task.errorInfor.Message);
            }
            else
            {
                MessageBox.Show("Sent Successfully");
            }
        };
        private interface IAsyncResultExt : IAsyncResult
        {
            Exception errorInfor { get;}
            bool IsCanceled { get;}
        }
    }
}

File: VB-Examples/02_SMTP Client/SendMail.vb
Imports OAuth
Imports Spire.Email
Imports Spire.Email.Smtp

Namespace SendMail
	Partial Public Class Form1
		Inherits Form
		Public Sub New()
			InitializeComponent()
		End Sub
		Private Sub btnRun_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnRun.Click
			If rBSync.Checked Then
				SendEmailSynchronously()
			ElseIf rBAsync.Checked Then
				SendEmailAsynchronously()
			End If
		End Sub
		Private Function GetSmtpClient() As SmtpClient
			'Connect outlook-Office using OAuth2.0 authentication
'               
'             *Please use TargetFrameworkVersion 4.5 or above to run this demo
'           
'             *Dependencies
'             AzureROPCTokenProvider.cs
'             AzureTokenResponse.cs
'             OAuthToken.cs
'             TokenType.cs
'             .\packages\Newtonsoft.Json.13.0.3\lib\net40\Newtonsoft.Json.dll
'             

			Dim host As String = Me.HostInput.Text
			Dim userName As String = Me.Username.Text
			Dim password As String = Me.Password.Text
			Dim port As UShort = UShort.Parse(Me.PortInput.Text)

			'Register your application with Azure AD to use Microsoft/Office365 OAUTH in your application, and getting the tenantId and clientId from the registered application.
			Dim tenantId As String = "9facb510-bbc6-4c95-a0f5-9a60d3315f3d"
			Dim clientId As String = "0a7976ae-ca80-4cb8-983d-9fea41e88dac"

			'scope: The scope to which the client is limited. Permissions for openid, offline_access and https://outlook.office.com/SMTP.Send.         
			Dim scopes() As String = { "openid", "email", "offline_access","https://outlook.office.com/SMTP.Send"}

			'Creating AzureROPCTokenProvider object to get an access token from a token server.
			Dim provider As New AzureROPCTokenProvider(tenantId, clientId, userName, password, scopes)
			Dim result As OAuthToken = provider.GetAccessToken()

			'Creating SmtpClient object to connect Outlook server
			Dim client As New SmtpClient(host, port, True, userName, result.Token, ConnectionProtocols.StartTls)
			client.Password = password

			Return client


			'Do not use OAuth2.0 authentication to connect to qq-email or Gmail
'            SmtpClient client = new SmtpClient();
'            client.Host = this.HostInput.Text;
'            client.Username = this.Username.Text ;
'            client.Password = this.Password.Text ;
'            client.Port = int.Parse(this.PortInput.Text);
'            client.ConnectionProtocols = ConnectionProtocols.Ssl;
'            return client;
		End Function
		Private Sub SendEmailSynchronously()
			Dim msg As New MailMessage(Me.FromAddress.Text,Me.ToAddress.Text)
			setContent(msg)
			Dim client As SmtpClient = GetSmtpClient()
			Try
				client.SendOne(msg)
				MessageBox.Show("Sent Successfully")
			Catch ex As Exception
				MessageBox.Show(ex.Message)
			End Try

		End Sub
		Private Sub SendEmailAsynchronously()
			Dim msg As New MailMessage(New MailAddress(Me.FromAddress.Text,Me.DisplayName.Text),Me.ToAddress.Text)
			setContent(msg)
			Dim client As SmtpClient = GetSmtpClient()
			client.BeginSendOne(msg, Callback)

		End Sub
		Private Sub setContent(ByVal msg As MailMessage)
			msg.Cc.Add(Me.CcInput.Text)
			msg.Bcc.Add(Me.BccInput.Text)
			msg.Subject = Me.SubjectInput.Text
			msg.BodyText = Me.TextInput.Text
		End Sub
		Dim Callback As AsyncCallback = New AsyncCallback(AddressOf A)
		Private Sub A(ByVal ar As IAsyncResult)
			Dim task As IAsyncResultExt = TryCast(ar, IAsyncResultExt)
			If task IsNot Nothing AndAlso task.IsCanceled Then
				MessageBox.Show("Cancel")
			End If
			If task IsNot Nothing AndAlso task.errorInfor IsNot Nothing Then
				MessageBox.Show("{0}", task.errorInfor.Message)
			Else
				MessageBox.Show("Sent Successfully")
			End If
		End Sub
		Private Interface IAsyncResultExt
			Inherits IAsyncResult
			ReadOnly Property errorInfor() As Exception
			ReadOnly Property IsCanceled() As Boolean
		End Interface
	End Class
End Namespace